texar/data/data/multi_aligned_data.py

```diff
@@ -0,0 +1,701 @@
+                #    'y_text_ids': [['1', '6', '10', '20', '2']],
+                #    'y_length': [5],
+                #    'z': [1000],
+                # }
+
+            ...
+
+            hparams={
+                'datasets': [
+                    {'files': 'd.txt', 'vocab_file': 'v.d', 'data_name': 'm'},
+                    {
+                        'files': 'd.tfrecord',
+                        'data_type': 'tf_record',
+                        "feature_original_types": {
+                            'image': ['tf.string', 'FixedLenFeature']
+                        },
+                        'image_options': {
+                            'image_feature_name': 'image',
+                            'resize_height': 512,
+                            'resize_width': 512,
+                        },
+                        'data_name': 't',
+                    }
+                ]
+                'batch_size': 1
+            }
+            data = MultiAlignedData(hparams)
+            iterator = DataIterator(data)
+            for batch in iterator:
+                # batch contains the following
+                # batch_ == {
+                #    'x_text': [['<BOS>', 'NewYork', 'City', 'Map', '<EOS>']],
+                #    'x_text_ids': [['1', '100', '80', '65', '2']],
+                #    'x_length': [5],
+                #
+                #    # "t_image" is a list of a "numpy.ndarray" image
+                #    # in this example. Its width equals to 512 and
+                #    # its height equals to 512.
+                #    't_image': [...]
+                # }
+
+    """
+
+    def __init__(self, hparams, device: Optional[torch.device] = None):
+        self._hparams = HParams(hparams, self.default_hparams())
+        self._other_transforms: List[
+            List[Callable[[Union[List[str], int, RawExample]], Example]]] = []
+        # Defaultizes hyperparameters of each dataset
+        datasets_hparams = self._hparams.datasets
+        defaultized_datasets_hparams = []
+        for ds_hpms in datasets_hparams:
+            data_type = ds_hpms.get("data_type", None)
+            defaultized_ds_hpms = HParams(ds_hpms,
+                                          _default_dataset_hparams(data_type))
+            defaultized_datasets_hparams.append(defaultized_ds_hpms)
+            self._other_transforms.append(
+                defaultized_ds_hpms.other_transformations)
+        self._hparams.datasets = defaultized_datasets_hparams
+
+        self._vocab = self.make_vocab(self._hparams.datasets)
+        self._embedding = self.make_embedding(self._hparams.datasets,
+                                              self._vocab)
+
+        self._dataset_features: List[Optional[Dict[str, Any]]] = []
+        self._name_prefix: List[str] = []
+        datasources: List[DataSource] = []
+        filters: List[Optional[Callable[[str], bool]]] = []
+        for _, hparams_i in enumerate(self._hparams.datasets):
```

Comment:
The logic below (and those in `process` and `collate`) have a great deal of duplicate code with the composed dataset classes (`MonoTextData`, `ScalarData`, `RecordData`). I guess the reason we need to copy their logic instead of using instances of those datasets are two fold:

1. `MonoTextData` does not provide interfaces to use shared vocabularies or embeddings.
2. Neither raw examples or processed examples can be directly accessed without constructing a `DataIterator` instance.

So my proposal is this:

- We create instances of the composed datasets as fully lazy and non-caching (so it does not access the data source nor store extra stuff), and use a null data source (e.g. `SequenceDataSource([])`).
- For `MonoTextData`, we add a private classmethod that constructs an instance of `MonoTextData` with possibly shared vocabulary & embeddings.
- In `process` and `collate`, we call the corresponding `process` and `collate` functions of each dataset.